File: stlc.py
from __future__ import annotations

import typing as ty
from typing import Any as TAny, Callable as TLam, List as TList, \
                   Dict as TDict

AbstractF = ty.TypeVar('AbstractF', bound=TLam[..., TAny])
def abstract(cls : AbstractF) -> TAny:
    class Abstract(cls): # type: ignore
        __slots__ = cls.__slots__
        def __init__(self : Abstract, *args : TAny, **kwargs : TAny) -> None:
            if type(self) == Abstract:
                raise TypeError(f"'{cls.__name__}' is an abstract class")
            super().__init__(*args ,**kwargs)
    return Abstract

@abstract
class TermI:
    __slots__ = ['_']

class Ann(TermI):
    __slots__ = ['e', 't']
    def __init__(self, expr : TermC, type_ : Type):
        super().__init__()
        self.e = expr
        self.t = type_
    def __repr__(self) -> str:
        return f"{self.e}::{self.t}"

class Bound(TermI):
    __slots__ = ['i']
    def __init__(self, pos : int):
        super().__init__()
        self.i = pos
    def __repr__(self) -> str:
        return f"(Bound {self.i})"

class Free(TermI):
    __slots__ = ['x']
    def __init__(self, name : Name):
        self.x = name
    def __repr__(self) -> str:
        return f"(Free {self.x})"

class App(TermI):
    __slots__ = ['e','ep']
    def __init__(self, termFun : TermI, termArg : TermC):
        self.e = termFun
        self.ep = termArg
    def __repr__(self) -> str:
        return f"{self.e}@{self.ep}"


@abstract
class TermC:
    __slots__ = ['_']
class Inf(TermC):
    __slots__ = ['i']
    def __init__(self, iterm : TermI):
        super().__init__()
        self.i = iterm
    def __repr__(self) -> str:
        return f"(Inf {self.i})"
class Lam(TermC):
    __slots__ = ['e']
    def __init__(self, expr : TermC):
        self.e = expr
    def __repr__(self) -> str:
        return f"(Lam {self.e})"

@abstract
class Name(): __slots__ = ['_']

class Global(Name):
    __slots__ = ['str']
    def __init__(self, str_ : str):
        self.str = str_
    def __repr__(self) -> str:
        return f"(Global {self.str})"
    def __eq__(self, other : Global) -> bool: # type: ignore
        assert(isinstance(other, Global))
        return self.str == other.str
    def __hash__(self) -> TAny:
        return self.str.__hash__()

class Local(Name):
    __slots__ = ['i']
    def __init__(self, pos : int):
        self.i = pos
    def __repr__(self) -> str:
        return f"(Local {self.i})"
    def __eq__(self, other : Local) -> bool: # type: ignore
        assert(isinstance(other, Local))
        return self.i == other.i
    def __hash__(self) -> TAny:
        return self.i.__hash__()

class Quote(Name):
    __slots__ = ['i']
    def __init__(self, pos : int):
        self.i = pos
    def __repr__(self) -> str:
        return f"(Quote {self.i})"
    def __eq__(self, other : Quote) -> bool: # type: ignore
        assert(isinstance(other, Quote))
        return self.i == other.i
    def __hash__(self) -> TAny:
        return self.i.__hash__()

@abstract
class Type: __slots__ = ['_']

class TFree(Type):
    __slots__ = ['x']
    def __init__(self, name : Name):
        self.x = name
    def __repr__(self) -> str:
        return f"(TFree {self.x})"
    def __eq__(self, other : TFree) -> bool: # type: ignore
        assert(isinstance(other, TFree))
        return self.x == other.x

class Fun(Type):
    __slots__ = ['t','tp']
    def __init__(self, typePrm : Type, typeRes : Type):
        self.t = typePrm
        self.tp = typeRes
    def __repr__(self) -> str:
        return f"(Fun {self.t} {self.tp})"
    def __eq__(self, other : Fun) -> bool: # type: ignore
        assert(isinstance(other, Fun))
        return self.t == other.tp and self.tp == other.tp

@abstract
class Value: __slots__=['_']

class VLam(Value):
    __slots__ = ['f']
    def __init__(self, lam : TLam[[Value], Value]):
        self.f = lam
    def __repr__(self) -> str:
        return f"(VLam {self.f})"

class VNeutral(Value):
    __slots__ = ['n']
    def __init__(self, neutral : Neutral):
        self.n = neutral
    def __repr__(self) -> str:
        return f"(VNeutral {self.n})"

@abstract
class Neutral: __slots__ = ['_']

class NFree(Neutral):
    __slots__ = ['x']
    def __init__(self, name : Name):
        self.x = name
    def __repr__(self) -> str:
        return f"(NFree {self.x})"
class NApp(Neutral):
    __slots__ = ['n', 'v']
    def __init__(self, neutral : Neutral, value : Value):
        self.n = neutral
        self.v = value
    def __repr__(self) -> str:
        return f"(NApp {self.n} {self.v})"

vfree : TLam[[Name],Value] = lambda n :  VNeutral(NFree(n))

Env = TList[Value]

def evalI(term : TermI, env : Env) -> Value:
    if isinstance(term, Ann):
        return evalC(term.e, env)
    elif isinstance(term, Free):
        return vfree(term.x)
    elif isinstance(term, Bound):
        return env[term.i]
    elif isinstance(term, App):
        return vapp(evalI(term.e, env), evalC(term.ep, env))
    raise TypeError(f"Unknown instance '{type(term)}'")

def vapp(v : Value, v1 : Value) -> Value:
    if isinstance(v, VLam):
        return v.f(v1)
    elif isinstance(v, VNeutral):
        return VNeutral(NApp(v.n, v1))
    raise TypeError(f"Unknown instance '{type(v)}'")

def evalC(term : TermC, env : Env) -> Value:
    if isinstance(term, Inf):
        return evalI(term.i, env)
    elif isinstance(term, Lam):
        lam_expr = term.e
        return VLam(lambda x : evalC(lam_expr, [x] + env))
    raise TypeError(f"Unknown instance '{type(term)}'")

@abstract
class Kind: __slots__ = ['_']
class Star(Kind):
    __slots__ = ['_']
    def __repr__(self) -> str:
        return f"Star"

@abstract
class Info: __slots__ = ['_']
class HasKind(Info):
    __slots__ = ['kind']
    def __init__(self, kind : Kind):
        self.kind = kind
    def __repr__(self) -> str:
        return f"(HasKind {self.kind})"
class HasType(Info):
    __slots__ = ['t']
    def __init__(self, type_ : Type):
        self.t = type_
    def __repr__(self) -> str:
        return f"(HasType {self.t})"

Context = TDict[Name, Info]

def kindC(c : Context, term : Type, k : Kind) -> None:
    assert(isinstance(k, Star))
    if isinstance(term, TFree):
        if isinstance(c[term.x], HasKind):
            return
        else:
            raise RuntimeError(f"unknown var identifier '{term.x}'")
    elif isinstance(term, Fun):
        kindC(c, term.t, Star())
        kindC(c, term.tp, Star())
        return
    raise TypeError(f"Unknown instance '{type(term)}'")

def typeI0(c : Context, term : TermI) -> Type:
    return typeI(0, c, term)

def typeI(i : int, c : Context, term : TermI) -> Type:
    if isinstance(term, Ann):
        kindC(c, term.t, Star())
        typeC(i, c, term.e, term.t)
        return term.t
    elif isinstance(term, Free):
        info = c[term.x]
        if isinstance(info, HasType):
            return info.t
        else:
            raise RuntimeError(f"unknown type identifier '{term.x}'")
    elif isinstance(term, App):
        f = typeI(i, c, term.e)
        if isinstance(f, Fun):
            typeC(i, c, term.ep, f.t)
            return f.tp
    raise TypeError(f"Unknown instance '{type(term)}'")

def dict_merge(a : TDict[TAny, TAny],
               b : TDict[TAny, TAny]) -> TDict[TAny, TAny]:
    a = a.copy()
    a.update(b)
    return a

def typeC(i : int, c: Context, term : TermC, ty : Type) -> None:
    if isinstance(term, Inf):
        ty1 = typeI(i, c, term.i)
        if ty != ty1:
            raise TypeError(f"type mismatch: {ty} != {ty1}")
        return
    elif isinstance(term, Lam) and isinstance(ty, Fun):
        typeC(i+1, dict_merge({Local(i): HasType(ty.t)}, c),
                   substC(0, Free(Local(i)), term.e), ty.tp)
        return
    raise TypeError(f"Type mismatch: term={term}', type={type}")

def substI(i : int, r : TermI, t : TermI) -> TermI:
    if isinstance(t, Ann):
        return Ann(substC(i,r,t.e), t.t)
    elif isinstance(t, Bound):
        return r if i == t.i else t
    elif isinstance(t, Free):
        return t
    elif isinstance(t, App):
        return substI(i, r, App(t.e, substC(i,r,t.ep)))
    raise TypeError(f"Unknown instance '{type(t)}'")

def substC(i : int, r : TermI, t : TermC) -> TermC:
    if isinstance(t, Inf):
        return Inf(substI(i,r,t.i))
    elif isinstance(t, Lam):
        return Lam(substC(i+1, r, t.e))
    raise TypeError(f"Unknown instance '{type(t)}'")

def quote0(v : Value) -> TermC:
    return quote(0,v)


"""
--------------------------------------------------
@claim
def quote1(_1 : int, _2 : Value) -> TermC: pass

generates:
quote1__dict = dict()
def quote1(_1 : int, _2 : Value) -> TermC:
    assert(isinstance(_1,int))
    assert(isinstance(_2,Value))
    quote1__args = (int, Value)
    return quote1_dict["_1__{int.__name__},
                        _2__{Value.__name__}"](_1, _2)

-----------------------------------------------------

@define
def quote1(i, _1 : VLam(f)):
    return Lam(quote(i+1, f(vfree(Quote(i)))))

generates

quote1__dict.update(f"_1__{int.__name__},
                      _2__{VLam.__name__}":quote1__VLam)
def quote1__VLam(_1 : int, _2 : VLam) -> TermC:
    assert(isinstance(_2, VLam))
    i = _1
    f = _2.expr
    return Lam(quote(i+1, f(vfree(Quote(i)))))

-------------------------------------------

@match
def quote1(i, _ : VNeutral(n)):
    return Inf(neutralQuote(i,n))

generates

quote1__dict.update(f"_1__{int.__name__},
                      _2__{NVeutral.__name__}":quote1__VNeutral)

def quote1__VNeutral(_1 : int, _2 : VNeutral) -> TermC:
    assert(isinstancE(_2, VNeutral))
    i = _1
    n = _2.neutral
    return Inf(Neutral(i,n))
"""

def quote(i : int, v : Value) -> TermC:
    if isinstance(v, VLam):
        return Lam(quote(i+1, v.f(vfree(Quote(i)))))
    elif isinstance(v, VNeutral):
        return Inf(neutralQuote(i,v.n))
    raise TypeError(f"Unknown instance '{type(v)}'")

def neutralQuote(i : int, n : Neutral) -> TermI:
    if isinstance(n, NFree):
        return boundfree(i,n.x)
    elif isinstance(n, NApp):
        return App(neutralQuote(i, n.n), quote(i,n.v))
    raise TypeError(f"Unknown instance '{type(n)}'")

def boundfree(i : int, x : Name) -> TermI:
    if isinstance(x, Quote):
        return Bound(i-x.i-1)
    else:
        return Free(x)

#e0 = quote0 (VLam (lambda x : VLam (lambda y : x)))
#print(e0)

id_ = Lam (Inf (Bound(0)))
const_ = Lam (Lam (Inf (Bound(1))))
tfree : TLam[[str], TFree] = lambda a : TFree (Global(a))
free : TLam[[str],Inf] = lambda x : Inf (Free (Global(x)))
term1 = App(Ann(id_,(Fun (tfree("a"),tfree("a")))), free("y"))
term2 = App(App(Ann(const_,Fun (Fun(tfree("b"),tfree("b")),
                   (Fun (tfree("a"),
                        (Fun (tfree("b"),tfree("b"))))))),
        id_), free("y"))
env1 : Context = {Global("y"): HasType(tfree("a")),
                  Global("a"): HasKind(Star())}
env2 = env1.copy()
env2.update({Global("b"): HasKind(Star())})

print(evalI(term1, []))
print(quote0(evalI(term1, [])))
print(quote0(evalI(term2, [])))
print(typeI0(env1, term1))
print(typeI0(env2, term2))

File: lampi/src/stlc.rs
type Int = i32;

pub fn test() -> Int {
    42
}

/*
#[derive(ADT)]
enum TermI {
    Ann(TermC, Type),
    Bound(Int),
    Free(Name),
    App(TermI, TermC),
};
 */

trait Boxed: Sized + Clone {
    fn b(self: &Self) -> Box<Self> {
        box self.clone()
    }
    fn dup(self: &Self) -> Self {
        self.clone()
    }
}
// ---------------------------------------------------------------------------

#[derive(Clone, Eq, PartialEq, Debug)]
pub enum TermI {
    Ann(Box<TermC>, Box<Type>),
    Bound(Int),
    Free(Box<Name>),
    App(Box<TermI>, Box<TermC>),
}
impl Boxed for TermI {}

// ---------------------------------------------------------------------------

#[derive(Clone, Eq, PartialEq, Debug)]
pub enum TermC {
    Inf(Box<TermI>),
    Lam(Box<TermC>),
}
impl Boxed for TermC {}

#[derive(Clone, Eq, PartialEq, Debug)]
pub enum Name {
    Global(String),
    Local(Int),
    Quote(Int),
}
impl Boxed for Name {}

// ---------------------------------------------------------------------------

#[derive(Clone, Eq, PartialEq, Debug)]
pub enum Type {
    TFree(Box<Name>),
    Fun(Box<Type>, Box<Type>),
}
impl Boxed for Type {}

// ---------------------------------------------------------------------------

use std::rc::Rc;
#[derive(Clone)]
pub enum Value {
    VLam(Rc<dyn Fn(&Value) -> Value>),
    VNeutral(Box<Neutral>),
}
impl Boxed for Value {}

impl std::fmt::Debug for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Value::VLam(_) => write!(f, "VLam(_)"),
            Value::VNeutral(b) => write!(f, "VNeutral({:?})", b),
        }
    }
}

// ---------------------------------------------------------------------------

#[derive(Clone, Debug)]
pub enum Neutral {
    NFree(Box<Name>),
    NApp(Box<Neutral>, Box<Value>),
}
impl Boxed for Neutral {}

// ---------------------------------------------------------------------------

fn vfree(n: Name) -> Value {
    Value::VNeutral(Neutral::NFree(n.b()).b())
}

#[derive(Clone, Eq, PartialEq, Debug)]
pub enum Kind {
    Star,
}
impl Boxed for Kind {}

// ---------------------------------------------------------------------------

#[derive(Clone, Eq, PartialEq, Debug)]
pub enum Info {
    HasKind(Box<Kind>),
    HasType(Box<Type>),
}
impl Boxed for Info {}

// ---------------------------------------------------------------------------

use std::collections::VecDeque;
pub type Env = VecDeque<Value>;

#[allow(non_snake_case)]
pub fn evalI(trm: &TermI, env: &Env) -> Value {
    match trm {
        TermI::Ann(e, _) => evalC(&e, env),
        TermI::Free(x) => vfree(x.dup()),
        &TermI::Bound(i) => env[i as usize].dup(),
        TermI::App(e, ep) => vapp(&evalI(&e, env), &evalC(&ep, env)),
    }
}

#[allow(non_snake_case)]
fn evalC(trm: &TermC, env: &Env) -> Value {
    match trm {
        TermC::Inf(i) => evalI(i, env),
        TermC::Lam(e) => {
            let env = env.clone();
            let e = e.dup();
            Value::VLam(Rc::new(move |x| {
                let mut env = env.clone();
                env.push_front(x.dup());
                evalC(&e, &env)
            }))
        }
    }
}

fn vapp(v1: &Value, v: &Value) -> Value {
    match v1 {
        Value::VLam(f) => f(v),
        Value::VNeutral(n) => Value::VNeutral(Neutral::NApp(n.b(), v.b()).b()),
    }
}

pub type Ctx = VecDeque<(Name, Info)>;
type Result<T> = std::result::Result<T, String>;

fn lookup<'a, 'b>(c: &'a Ctx, n: &'b Name) -> Option<&'a Info> {
    if let Some((_, i)) = c.iter().find(|x| x.0 == *n) {
        Some(i)
    } else {
        None
    }
}

#[allow(non_snake_case)]
fn kindC(ctx: &Ctx, t: &Type, k: &Kind) -> Result<()> {
    match (t, k) {
        (Type::TFree(x), Kind::Star) => match lookup(ctx, x) {
            Some(Info::HasKind(box Kind::Star)) => Ok(()),
            None => Err(format!("unk var identifier {:?}", x)),
            _ => panic!("unhandled case {:?}", x),
        },
        (Type::Fun(k, k1), Kind::Star) => {
            kindC(ctx, k, &Kind::Star)?;
            kindC(ctx, k1, &Kind::Star)
        }
    }
}

#[allow(non_snake_case)]
pub fn typeI0(c: &Ctx, term: &TermI) -> Result<Type> {
    typeI(0, c, term)
}

#[allow(non_snake_case)]
fn typeI(i: Int, c: &Ctx, trm: &TermI) -> Result<Type> {
    match trm {
        TermI::Ann(e, t) => {
            kindC(c, &t, &Kind::Star)?;
            typeC(i, c, &e, &t)?;
            Ok(t.dup())
        }
        TermI::Free(x) => {
            if let Some(x) = lookup(c, &x) {
                match x {
                    Info::HasType(t) => Ok(t.dup()),
                    _ => panic!("unhandled case {:?}", x),
                }
            } else {
                Err(format!("unk type identifier {:?}", x))
            }
        }
        TermI::App(e, ep) => {
            let s = typeI(i, c, &e)?;
            match s {
                Type::Fun(t, tp) => {
                    typeC(i, c, &ep, &t)?;
                    Ok(tp.dup())
                }
                _ => Err(format!(" illegal application {:?}", (e, ep))),
            }
        }
        _ => panic!("unhandled case {:?}", trm),
    }
}

#[allow(non_snake_case)]
fn typeC(i: Int, c: &Ctx, term: &TermC, typ: &Type) -> Result<()> {
    match (term, typ) {
        (TermC::Inf(e), _) => {
            let tp = &typeI(i, c, e)?;
            if typ != tp {
                return Err(format!("[inf] type mismatch {:?}", (typ, tp)));
            }
            Ok(())
        }
        (TermC::Lam(e), Type::Fun(t, tp)) => {
            let mut c = c.clone();
            c.push_front((Name::Local(i), Info::HasType(t.b())));
            let s = substC(0, &TermI::Free(Name::Local(i).b()), e);
            typeC(i + 1, &c, &s, tp)
        }
        _ => Err(format!("oops, type mismatch {:?}", (term, typ))),
    }
}

#[allow(non_snake_case)]
fn substI(i: Int, r: &TermI, t: &TermI) -> TermI {
    match t {
        TermI::Ann(e, t) => TermI::Ann(substC(i, r, e).b(), t.b()),
        &TermI::Bound(j) => {
            if i == j {
                r.dup()
            } else {
                TermI::Bound(j)
            }
        }
        TermI::Free(y) => TermI::Free(y.b()),
        TermI::App(e, ep) => TermI::App(substI(i, r, &e).b(), substC(i, r, &ep).b()),
    }
}

#[allow(non_snake_case)]
fn substC(i: Int, r: &TermI, t: &TermC) -> TermC {
    match t {
        TermC::Inf(e) => TermC::Inf(substI(i, r, e).b()),
        TermC::Lam(e) => TermC::Lam(substC(i + 1, r, e).b()),
    }
}

pub fn quote0(v: &Value) -> TermC {
    quote(0, v)
}

fn quote(i: Int, v: &Value) -> TermC {
    match v {
        Value::VLam(f) => TermC::Lam(quote(i + 1, &f(&vfree(Name::Quote(i)))).b()),
        Value::VNeutral(n) => TermC::Inf(neutralQuote(i, n).b()),
    }
}

#[allow(non_snake_case)]
fn neutralQuote(i: Int, n: &Neutral) -> TermI {
    match n {
        Neutral::NFree(x) => boundfree(i, x),
        Neutral::NApp(n, v) => TermI::App(neutralQuote(i, n).b(), quote(i, v).b()),
    }
}

fn boundfree(i: Int, n: &Name) -> TermI {
    match n {
        Name::Quote(k) => TermI::Bound(i - k - 1),
        _ => TermI::Free(n.b()),
    }
}
